server/websocket.go

```diff
@@ -1295,7 +1296,13 @@ func (c *client) wsCollapsePtoNB() (net.Buffers, int64) {
 			mfs = 0
 		}
 		buf := bytes.NewBuffer(nbPoolGet(usz))
-		cp, _ := flate.NewWriter(buf, flate.BestSpeed)
+		cp := c.ws.compressor
+		if cp == nil {
+			c.ws.compressor, _ = flate.NewWriter(buf, flate.BestSpeed)
+			cp = c.ws.compressor
+		} else {
+			cp.Reset(buf)
+		}
```

Comment:
The issue with this is that we hold onto the compressor and hence the bytes.Buffer and hence the []byte underneath.

Reply:
The compressor will be held yes, as will the `bytes.Buffer` (which is only a few bytes holding the underlying slice reference), but the actual underlying `[]byte` itself will not be held indefinitely.

Now that we aren't copying from that buffer anymore, the underlying memory is reused and queued up in `queueOutbound` and returned to the pool when either `flushOutbound` or `flushAndClose` completes. On the next pass, we reset the compressor with a new `bytes.Buffer` with new underlying memory from the pool.

Reply:
But unless we do a reset on the compressor it will hold a reference all the way down to the byte slice.

Reply:
The reference is already held anyway by the `queueOutbound` because we're no longer copying, but I've added a `cp.Reset(nil)` to force it to drop the reference as we always end up calling `cp.Reset(buf)` before the next use anyway. That way the GC should be able to do its work post-flush rather than having to wait for the next write.